### ISSUE TITLE
CmdStanMCMC object - add function `get_var` and `get_vars`

### ISSUE THREAD

**Issue by mitzimorris** [stan-dev/cmdstanr#183]
For container program variables, i.e., vectors, matrices, and arrays, users need to be able to get the draws for that variable structured accordingly.  For scalar variables, this is easy; it is just one column's worth of data and the variable name matches the column name.
For structured variables, we need to assemble all element of the container.

In RStan this is done by the `extract` function.

The name `extract` is a poor choice, `extract` is associated with SQL and there is an `extract` function in the tidyverse, also in corresponding Python libraries.

Possible names are `get_param`, `get_var` or `get_variable`.

There is a parallel issue on CmdStanPy - https://github.com/stan-dev/cmdstanpy/issues/245.

Let's use this issue to discuss names.

**Comment by mitzimorris** [comment]
https://github.com/stan-dev/cmdstanpy/pull/252

**Comment by rok-cesnovar** [comment]
@jgabry the changes in https://github.com/stan-dev/posterior/pull/106 are going to allow us to implement this right? Or am I misunderstanding that?

**Comment by mitzimorris** [comment]
just to be clear, names we ended up going with are: `stan_variable`, `stan_variables` and `sampler_diagnostics` (for the sampler parameters `lp__` et al).

**Comment by jgabry** [comment]
> @jgabry the changes in [stan-dev/posterior#106](https://github.com/stan-dev/posterior/pull/106) are going to allow us to implement this right? Or am I misunderstanding that?

Yeah, I'm pretty sure it should let us do something similar to what @mitzimorris did.

**Comment by mitzimorris** [comment]
PR https://github.com/stan-dev/cmdstanpy/pull/346 updated functionality of `stan_variable` et al, also added properties `num_draws_warmup`, `num_draws_sampling`, `thin`, `stan_vars_dims`, `stan_vars_cols`, and `sampler_vars_cols`  which are used by downstream processing to access rows, columns of the draws array.

**Comment by rok-cesnovar** [comment]
Just curious, why is it num_draws_warmup/sampling? Doesnt that differ from the argument naming style (iter_*)?

cmdstanr will solve this one with rvar from posterior which will be available soon.

**Comment by mitzimorris** [comment]
> why is it num_draws_warmup/sampling? Doesnt that differ from the argument naming style (iter_*)

big difference:  https://cmdstanpy.readthedocs.io/en/v0.9.68/sample.html

> Note that the terms iterations and draws are not synonymous. The HMC sampler is configured to run a specified number of iterations. By default, at the end of each iteration, the values of all sampler and parameter variables are written to the Stan CSV output file. Each reported set of estimates constitutes one row’s worth of data, each row of data is called a “draw”. The sampler argument thin controls the rate at which iterations are recorded as draws. By default, thin is 1, so every iteration is recorded. Increasing the thinning rate will reduce the frequency with which the iterations are recorded, e.g., thin = 5 will record every 5th iteration.

**Comment by mitzimorris** [comment]
> cmdstanr will solve this one with rvar from posterior which will be available soon.

will it be exposed with similar function names, or are interfaces diverging along R-thonic / Py-thonic lines?

**Comment by jgabry** [comment]
> PR stan-dev/cmdstanpy#346 updated functionality of stan_variable et al, also added properties num_draws_warmup, num_draws_sampling, thin, stan_vars_dims, stan_vars_cols, and sampler_vars_cols which are used by downstream processing to access rows, columns of the draws array.

> > cmdstanr will solve this one with rvar from posterior which will be available soon.
> 
> will it be exposed with similar function names, or are interfaces diverging along R-thonic / Py-thonic lines?

I'm now thinking that this is one place where we'll diverge because in R we're just outsourcing this to the posterior package. I think it makes sense for CmdStanPy to have these methods and additional properties because there's no posterior package in python, but I think adding all this to CmdStanR would just be duplicating a lot of stuff already provided by posterior, which seems like overkill / extra code to maintain with no great benefit. 

So instead of needing to address this issue for v1.0, we just need to fix https://github.com/stan-dev/cmdstanr/issues/582 so it's clear how the user can get the correct rvars objects. 

We can definitely reopen this if we decide to reconsider.

**Comment by mitzimorris** [comment]
here's the user's question:

"I know how to get the structured variable in RStan.  given Stan program:

```
matrix[M, N] a;
```

given fitted object `foo`, which contains a total of D draws (from I chains of J iterations each)
in RStan,  I can use the extract function to get "a" as an M x N x D object.
how do I do this in Cmdstanr?"


I don't see this answered clearly anywhere in the CmdStanR documentation.

**Comment by bob-carpenter** [comment]
@mitzimorris:  You can convert cmdstanr output to rstan output then using rstan's extract.

@jgabry:  Is there a way to do this directly in cmdstanr or through posterior?  If not, copying over the rstan implementation would make sense as then users could extract in cmdstanr without having to install rstan.

**Comment by mitzimorris** [comment]
this needs to be implemented in CmdStanR.

depending on posterior is not good enough
 - minimal documentation
 - no way to get simple data values - everything is run through dplyr, results are tibbles and seem to have "pillar" formatting attached - I've wasted hours trying to make sure that what I'm seeing here:
<img width="384" alt="Untitled 2" src="https://github.com/stan-dev/cmdstanr/assets/5466612/8c88b13e-dbf3-49aa-8b5a-15599f4523bb">
is just the result of bad printing to the RStudio console and not some problem with my data or model.

**Comment by bob-carpenter** [comment]
This issue has been closed @mitzimorris.  I suggest either re-opening this issue or starting a new one with a concrete feature proposal.

**Comment by mitzimorris** [comment]
thanks for the clear explanation of how to do this!   definitely should go into the docs.

> Though it takes 3 lines instead of 1. And I guess the draws_of() function is a bit obscure (I didn't know about it until now)

exactly.  for Andrew, and just about every other user, 3 lines is alot and the `draws_of()` function is totally unknown.

**Comment by jgabry** [comment]
This will now be better documented in the CmdStanR vignettes: https://github.com/stan-dev/cmdstanr/pull/955/files
